Set up directories for migrated Dart 2.0 tests and migrate a couple.

munificent opened the pull request:
- Create new "_2" directories where tests that have been validated as
  ready for Dart 2.0 will end up.
- Create empty status files for each directory.
- Add those directories to the set of default selectors you get when you
  run test.py. This gets the VM bots running them.
- Get the DDC bots running those suites.
- Move abstract_exact_selector_test over to dart2js since it's a
  dart2js-specific regression test and not an actual language test.
- Migrate corelib/apply_test.dart.
- Delete abstract_beats_arguments[2]_test.dart since that code is
  statically wrong in 2.0 and can't be run.

This doesn't get the dart2js bots running the new suites. I'll email
the relevant folks to get help with that.

munificent committed:
Set up directories for migrated Dart 2.0 tests and migrate a couple.

- Create new "_2" directories where tests that have been validated as
  ready for Dart 2.0 will end up.
- Create empty status files for each directory.
- Add those directories to the set of default selectors you get when you
  run test.py. This gets the VM bots running them.
- Get the DDC bots running those suites.
- Move abstract_exact_selector_test over to dart2js since it's a
  dart2js-specific regression test and not an actual language test.
- Migrate corelib/apply_test.dart.
- Delete abstract_beats_arguments[2]_test.dart since that code is
  statically wrong in 2.0 and can't be run.

This doesn't get the dart2js bots running the new suites. I'll email
the relevant folks to get help with that.

jcollins-g commented on tests/language/abstract_beats_arguments2_test.dart:
```diff
@@ -1,14 +0,0 @@
-// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
```
This being dropped because it is testing a static type warning and that is covered elsewhere?

munificent replied:
Yes. In 2.0, a program that tries to construct an abstract class is statically erroneous.

jcollins-g replied:
lgtm

jcollins-g commented on tests/language/abstract_beats_arguments2_test.dart:
```diff
@@ -1,14 +0,0 @@
-
-import "package:expect/expect.dart";
-
-abstract //# 01: static type warning
```
we're deleting this because it is covered elsewhere?

munificent commented:
@stereotype441, this deletes a couple of tests of the behavior around constructing an abstract class. I see that analyzer/test/generated/static_type_warning_code_test.dart has a test for that error. Are you OK with killing the two language tests, or should I do something else with them?

munificent committed:
Resurrect abstract_beats_arguments_test.dart.

It usefully checked that an implementation reports an warning (now
error) if you construct an abstract class.

Also added support to test.dart to mark a test as expecting to
generate a compile error in the test itself. That way, the status file
reflects what is *wrong* about the current status, not what is *right*.

munificent commented:
OK, after talking to @stereotype441, I made some changes. The abstract_beats_arguments_test.dart test is migrated over to language_2 and all of the code not related to generating the static warning (now error) is removed. So it's now a pure test that an implementation reports a static error if you construct an abstract class.

To manage that in test.dart and the status files, I allow the `@compile-error` annotation we use in the co19 tests so that the test itself can say "I should generate a compile error". That way, in the status file, a correct implementation has no entry for the file. An incorrect implementation (the VM and dart2js today) has:

```
abstract_beats_arguments_test: MissingCompileTimeError
```

I think that reads about right. Better, of course, would be a system like front_end has where we can associate the error with a specific location in the source and the test runner can validate that. I'd like to get there, but we won't have that ready in time for the migration so, for now, I think we can get by with just validating that the test reports an error at all.

@peter-ahe-google, what do you think of allowing `@compile-error` in non-co19 files for now?

stereotype441 commented on tests/corelib_2/symbol_map_helper.dart:
```diff
@@ -0,0 +1,15 @@
+library dart.test.symbol_map_helper;
+
+// TODO(ahe): Update map literals to avoid this method.
+Map<Symbol, dynamic> symbolMapToStringMap(Map<String, dynamic> map) {
```
I realize you just copied this code from tests/corelib, but shouldn't this be called `stringMapToSymbolMap`?

munificent replied:
🤷‍♂️ I tried to do the merge mostly mechanically. I think there are other yet-to-be-migrated tests that use this library, so it might make it harder to migrate those if I rename this now.

stereotype441 replied:
Fair enough.  Cleaning up stuff like this is orthogonal to the migration anyhow :)

stereotype441 reviewed (approved):
lgtm

peter-ahe-google commented:
> what do you think of allowing @compile-error in non-co19 files for now?

I don't think that's a good approach. The problem is that people write a test, then something changes in the the language and the test ends up having a different compile-time error. I've found a lot of these kinds of tests as I've been implementing Fasta, so this isn't hypothetical.

Ideally, we'll be able to pinpoint each compile-time error somewhat precisely, but until we have a shared front-end, we can't expect all implementations to report errors in the same way.

Could you set up a GVC so we can discuss the various options?

stereotype441 commented on tests/language_2/abstract_beats_arguments_test.dart:
```diff
@@ -0,0 +1,16 @@
+}
+
+void main() {
+  /* @compile-error */ new A();
```
If you change this to:
```dart
void main() {
  /*@compile-error=unspecified*/ new A();
}
```

Then it will be compatible with the annotation framework I'm already using for type inference tests, and we can replace `unspecified` with the appropriate error code once we know what it is.  (Note: make sure there is no space between `*` and `@`).

munificent replied:
Dig it. Done.

munificent committed:
Change static error syntax to match front_end notation.

stereotype441 commented:
Changes as of e765836cede2fc7b772c935ef0c8e13270fdce82 lgtm.

munificent committed:
Migrate abstract_getter_test.

munificent committed:
Merge branch 'master' into 2-tests
